app/jobs/autoscaling.py

```diff
@@ -181,26 +215,166 @@ def clamp_to_valid_range(asg, desired_capacity):
+
+
+def instances_to_drain(asg_instance_name, draining_tag, num_instances):
+    cmd = "aws ec2 describe-instances --filters 'Name=tag:Name,Values={asg_instance_name}' --query 'Reservations[*].Instances[*].[InstanceId,LaunchTime,Tags]' --no-paginate"
+    cmd = cmd.format(asg_instance_name=asg_instance_name)
+    aws_response = json.loads(aws_command(cmd))
+    zipped_instance_ids_and_launch_times = [(instance[0], iso2unixtime(instance[1]))
+        for instance_list in aws_response
+        for instance in instance_list
+        if draining_tag not in [tag["Key"] for tag in instance[2]]
+    ]
+    instance_ids_sorted_by_increasing_launch_time = [item[0] for item in sorted(zipped_instance_ids_and_launch_times, key=itemgetter(1))]
+    return instance_ids_sorted_by_increasing_launch_time[:num_instances]
+
+
+def instances_to_rescue(asg, tag_list, draining_tag, num_instances):
+    '''
+    Rescue the instances that have been draining for the least amount of time.
+    That way we can probably terminate the non-rescued instances sooner.
+    '''
+    draining_instances = get_draining_servers(asg, tag_list, draining_tag)
+    draining_instances_sorted = [key for key, _value in sorted(draining_instances.items(), key=itemgetter(1), reverse=True)]
+    return draining_instances_sorted[:num_instances]
+
+
+def get_draining_servers(asg, tag_list, draining_tag):
+    ''' returns a map of draining instance IDs to the time they started draining '''
+    protected_instance_ids = [inst["InstanceId"] for inst in asg["Instances"] if inst["ProtectedFromScaleIn"]]
+    instance_dict = { item['ResourceId']: int(item['Value']) for item in tag_list if item['Key'] == draining_tag and item['ResourceId'] in protected_instance_ids }
+    return instance_dict
+
+
+def count_running_alignment_jobs(asg, tag_list, job_tag_prefix, job_tag_expiration):
+    ''' returns a map of instance IDs to number of jobs running on the instance '''
+    instance_ids = instances_in(asg)
+    count_dict = { id: 0 for id in instance_ids }
+    expired_jobs = []
+    for item in tag_list:
+        if item['Key'].startswith(job_tag_prefix):
+            job_tag = item['Key']
+            instance_id = item['ResourceId']
+            unixtime = int(item['Value'])
+            if unixtime_now() - unixtime < job_tag_expiration:
+                count_dict[instance_id] += 1
+            else:
+                expired_jobs.append(job_tag)
+    if expired_jobs:
+        print "The following job tags have expired and will be deleted: " + ", ".join(expired_jobs)
+        delete_tags_from_instances(instance_ids, expired_jobs)
+    return count_dict
+
+
+def delete_tags_from_instances(instance_ids, tag_keys):
+    if not instance_ids or not tag_keys:
+        return
+    cmd = "aws ec2 delete-tags --resources {list_instances} --tags {list_tags}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), list_tags=' '.join(['Key='+tag for tag in tag_keys]))
+    aws_command(cmd)
```

Comment:
If this operation fails, will it be repeated next time autoscaling is reincarnated?

Reply:
Yes: if the the old tag still exists, it will be seen by describe-tags and cleaned up in the next incarnation